File: combinatorics/resource/ST/project_5/bai_toan_4/mutil_graph/al_to_eal.cpp
#include <iostream>
#include <vector>
#include <map>
#include <sstream>

using namespace std;

vector<tuple<int, int, int>> alToEal(const vector<vector<int>>& adjList) {
    vector<tuple<int, int, int>> eal;
    for (int u = 0; u < adjList.size(); ++u) {
        map<int, int> freq;
        for (int v : adjList[u]) {
            freq[v]++;  // Dem so lan xuat hien cua v
        }
        for (const auto& [v, w] : freq) {
            eal.push_back({u, v, w});  // Tao bo (u,v,w)
        }
    }
    return eal;
}

int main() {
    int n;
    cout << "Nhap so dinh: ";
    cin >> n;

    vector<vector<int>> adjList(n);
    cin.ignore(); // Xoa bo dem

    cout << "Nhap danh sach ke:\n";
    for (int i = 0; i < n; ++i) {
        cout << "Ke cua dinh " << i << ": ";
        string line;
        getline(cin, line);
        istringstream iss(line);
        int x;
        while (iss >> x)
            adjList[i].push_back(x);
    }

    // Chuyen doi va in ket qua
    auto eal = alToEal(adjList);
    cout << "Danh sach ke mo rong (EAL):\n";
    for (const auto& [u, v, w] : eal) {
        cout << "(" << u << ", " << v << ", " << w << ")\n";
    }

    return 0;
}

File: combinatorics/resource/ST/project_5/bai_toan_4/mutil_graph/al_to_eal.py
from collections import Counter

def al_to_eal(adj_list):
    eal = []
    for u, neighbors in enumerate(adj_list):
        freq = Counter(neighbors)  # Dem so lan moi dinh xuat hien
        for v, w in freq.items():
            eal.append((u, v, w))
    return eal

# Nhap tu ban phim
n = int(input("Nhap so dinh: "))
adj_list = []

print("Nhap danh sach ke:")
for i in range(n):
    row = list(map(int, input(f"Ke cua dinh {i}: ").split()))
    adj_list.append(row)

# Chuyen doi va in ket qua
eal = al_to_eal(adj_list)
print("Danh sach ke mo rong (EAL):")
for u, v, w in eal:
    print(f"({u}, {v}, {w})")